jwst/extract_1d/soss_extract/soss_extract.py

```diff
@@ -1508,11 +1561,57 @@ def run_extract1d(
     # Run the first integration to get the Tikhonov factor for the rest
     scidata = cube_model.data[0].astype("float64")
     scierr = cube_model.err[0].astype("float64")
-    scimask = np.bitwise_and(cube_model.dq[0], dqflags.pixel["DO_NOT_USE"]).astype(bool)
+    # Only mask pixels that are bad in *all* integrations
+    scimask = np.all(cube_model.dq & dqflags.pixel["DO_NOT_USE"] != 0, axis=0)
+
     refmask = bitfield_to_boolean_mask(
         cube_model.dq[0], ignore_flags=dqflags.pixel["REFERENCE_PIXEL"], flip_bits=True
     )
 
+    # Pixels that are bad in all integrations
+
+    allbad = np.sum(np.isfinite(cube_model.err) &
+                    np.isfinite(cube_model.data), axis=0) == 0
+
+    # Make a copy of the cube model where bad pixels in any individual
+    # integration are replaced with a running median in time.
+
+    data_nanreplaced = cube_model.data*1.
+    data_infilled = cube_model.data*1.
+
+    # We will use themedian uncertainty throughout the calculation.
+    # Use of the median uncertainty means that the matrices used in
+    # the ATOCA algorithm are shared between all integrations.
+
+    mederr = np.nanmedian(cube_model.err*1., axis=0)
+    meddata = np.nanmedian(data_nanreplaced, axis=0)
+
+    mederr[allbad] = np.inf
+    mederr[refmask] = np.inf
+    meddata[allbad] = 0
+
+    # Fill in values for pixels that are bad in individual integrations
+
+    for i in range(data_nanreplaced.shape[0]):
+        indx = ~np.isfinite(data_nanreplaced[i])
+        data_infilled[i][indx] = meddata[indx]
+
+    # Apply a median filter in time to make the final replacements of
+    # bad pixels in individual integrations.  Use a running median of
+    # 9 time steps for now.
+
+    medfilt = ndimage.median_filter(data_infilled, 9, axes=(0))
+
+    for i in range(data_nanreplaced.shape[0]):
+        indx = ~np.isfinite(data_nanreplaced[i])
+        data_nanreplaced[i][indx] = medfilt[i][indx]
+
+    # For the first integration, we will use the data with no NaNs
+    # and we will use the median uncertainties.
+
+    scidata = data_nanreplaced[0]
```

Comment:
Instead of processing the first integration to estimate the best regularization factor and wavelength grid, which is what's currently done, could you apply `process_one_integration` to the median data, and then process all integrations according to the result of that?

It appears that doing so would nicely fix a known issue, which is that if your first integration (for whatever reason) has lots of bad pixels and is unrepresentative of the rest of the integrations, then you might end up using and unrepresentative regularization factor and/or wave grid.

Reply:
This is a good thought.  It would change the behavior, because one integration has more noise than a median.  I think we have two options that we should explore:
1. Adopt your suggestion as-is.
2. Estimate the per-pixel noise and add a Gaussian realization of this.  This would be subtle: the per-pixel noise should be estimated from a robust measurement of the scatter in the difference between successive integrations.

We can do some checks before meeting on December 8.  It is subtle in the sense that, if the flux is constant, the noise will get beaten down in the median.  If the flux changes by a lot compared to the noise, the noise will not be smaller in the median.